### PR TITLE
[SPARK-39497][SQL] Improve the analysis exception of missing map key column

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Sometimes users forgot to add single quotes on the map key string literal, for example `map_col[a]`. In such a case, the Analyzer will throw an exception:
```
[MISSING_COLUMN] Column 'struct.a' does not exist. Did you mean one of the following? ...
```
We can improve this message by saying that the user should append single quotes if the map key is a string literal.

```
[UNRESOLVED_MAP_KEY] Cannot resolve column 'a' as a map key. If the key is a string literal, please add single quotes around it. Otherwise, did you mean one of the following column(s)? ...
```

### Why are the changes needed?

Error message improvement

### Does this PR introduce _any_ user-facing change?

Yes but trivial, an improvement on the error message of unresolved map key column

### How was this patch tested?

New UT

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @srielau as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 1]
```diff
@@ -160,28 +181,28 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
...
           case hof: HigherOrderFunction
               if hof.argumentsResolved && hof.checkArgumentDataTypes().isFailure =>
             hof.checkArgumentDataTypes() match {
               case TypeCheckResult.TypeCheckFailure(message) =>
                 hof.failAnalysis(
                   s"cannot resolve '${hof.sql}' due to argument data type mismatch: $message")
             }
+
+          // If an attribute can't be resolved as a map key of string type, either the key should be
+          // surrounded with single quotes, or there is a typo in the attribute name.
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:198]: I tried to match `element_at` and `try_element_at` here. But before the functions are not resolved yet in the rule `ResolveFunctions`, the error `MISSING_COLUMN` is thrown.
Checking the unresolved function name and arguments can work but is quite ugly. I decided to handle the `[]` operator first.

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala [hunk 2]
```diff
@@ -401,6 +401,24 @@ class QueryCompilationErrorsSuite
     )
   }
 
+  test("UNRESOLVED_MAP_KEY: string type literal should be quoted") {
+    checkAnswer(sql("select m['a'] from (select map('a', 'b') as m, 'aa' as aa)"), Row("b"))
+    checkErrorClass(
+      exception = intercept[AnalysisException] {
+        sql("select m[a] from (select map('a', 'b') as m, 'aa' as aa)")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala:408]: In `CheckAnalysis`, how can we distinguish `m.a` and `m[a]`?

Comment (gengliangwang) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryCompilationErrorsSuite.scala:408]: m.a is equivalent to m['a'] so it won't fail analysis

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -91,6 +91,27 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog {
...
+  private def failUnresolvedAttribute(
+      operator: LogicalPlan,
+      a: Attribute,
+      errorClass: String): Nothing = {
+    val unresolvedAttribute = a.sql
+    val candidates = operator.inputSet.toSeq.map(_.qualifiedName)
+    val orderedCandidates = StringUtils.orderStringsBySimilarity(unresolvedAttribute, candidates)
+    a.failAnalysis(
+      errorClass = errorClass,
+      messageParameters = Array(unresolvedAttribute, orderedCandidates.mkString(", ")))
```
Comment (srielau) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:112]: Both candidates and unresolvedAttribute should be wrapped with toSQLId

**Review by srielau — APPROVED** [review]
Added a comment about toSQLId

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala (hunk 3 above)
Comment (gengliangwang) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:112]: I see. I will update this one after https://github.com/apache/spark/pull/36891 is merged.

**Comment by gengliangwang** [comment]
Merging to master

**Pull request closed without merging**